(fix) - Tooltip appears for button when focus not on button, Does not get dismissed by clicking outside

gavin-everett-genesys opened the pull request:
Ticket : https://inindca.atlassian.net/browse/COMUI-1103

**Description :** 
When gux-tooltip is added to a gux-button, when we click on that button and either

The button gets programmatically disabled , or,

we follow up with a click outside the button.

we get an issue that the tooltip still appears for that button, and cannot be dismissed by clicking anywhere outside the button. The tooltip can only be disabled by hovering over that button again.

**Res:**
I removed the setTimeout/clearTimeout for the css transition as I dont think it is needed. Instead I updated the transition property to 1s in the css to align with the spark design. I also added a disabled check to the **mouseenter** as when the button or element is disabled on hover it will still display the tooltip. I just check to see if that element has an attribute of disabled if so the tooltip will not be displayed.

katie-bobbe-genesys reviewed (changes requested):
This changes the appearance of how the tooltip appears on hover. Instead of a delay of 750ms then a 250ms fade in, the tooltip slowly fades in over 1s. Is there a way to address this bug while preserving the original fade in behavior of the tooltip?

gavin-everett-genesys commented:
> This changes the appearance of how the tooltip appears on hover. Instead of a delay of 750ms then a 250ms fade in, the tooltip slowly fades in over 1s. Is there a way to address this bug while preserving the original fade in behavior of the tooltip?

I think there is a `transition-delay` property. I will test this out.

gavin-everett-genesys commented:
Added in `pointerenter`/`pointerleave` in replacement for `mouseenter`/`mouseleave` . The `pointerenter` still executes the tooltip on hover even when the element is `disabled`. I was thinking that I will need conditional logic to not execute the show method if the disabled attribute is present so i added in a ternary to check if it is present. I also was testing out `mouseover` and `mouseout` which do not show the tooltip on hover works in all browsers except for firefox.

For the transition instead of using the `setTimeout` in javascript there is a property called `transition-delay` that will delay the transition for a set amount of time. I set this to 750ms to match the initial implementation. I think it matches the spark design. I think this may be a cleaner way then a setTimeout but open to changes.

Link to spark design : https://spark.genesys.com/7978beca0/p/366caf-tooltip

katie-bobbe-genesys commented:
I found that the animation only works the first time you hover over the target element
[![Image from Gyazo](https://i.gyazo.com/bb75ac890ab45b4ae027781221a07b1c.gif)](https://gyazo.com/bb75ac890ab45b4ae027781221a07b1c)
I'm also finding that this change removes the tooltip animation for any component that uses the `gux-tooltip-title` internally, like `gux-tabs` and `gux-tag`. I think using `pointerenter` instead of `mouseover` is creating this breaking change.

katie-bobbe-genesys commented on src/components/stable/gux-tooltip/gux-tooltip.tsx:
```diff
@@ -24,10 +24,10 @@ import { GuxTooltipPlacements } from './gux-tooltip.types';
-  private mouseenterHandler: () => void = () => this.show();
-  private mouseleaveHandler: () => void = () => this.hide();
+  private pointerenterHandler: () => void = () =>
+    !this.forElement.hasAttribute('disabled') ? this.show() : null;
```
Wouldn't there be situations where you would want to see the tooltip on a disabled element? For example, if a disabled element has truncated text you might want to see the full text on hover.

daragh-king-genesys replied:
I agree. The tooltip showing when a button is disabled is not a bug for me.

katie-bobbe-genesys commented:
> I found that the animation only works the first time you hover over the target element [![Image from Gyazo](https://camo.githubusercontent.com/2e387bb348dffbeebc26c07ad02710a612ee664029dcfdc5ce7fd6184d6cd43b/68747470733a2f2f692e6779617a6f2e636f6d2f62623735616338393061623435623461653032373738313232316130376231632e676966)](https://gyazo.com/bb75ac890ab45b4ae027781221a07b1c) [ ![Image from Gyazo](https://camo.githubusercontent.com/2e387bb348dffbeebc26c07ad02710a612ee664029dcfdc5ce7fd6184d6cd43b/68747470733a2f2f692e6779617a6f2e636f6d2f62623735616338393061623435623461653032373738313232316130376231632e676966) ](https://gyazo.com/bb75ac890ab45b4ae027781221a07b1c) [ ](https://gyazo.com/bb75ac890ab45b4ae027781221a07b1c) I'm also finding that this change removes the tooltip animation for any component that uses the `gux-tooltip-title` internally, like `gux-tabs` and `gux-tag`.

To fix the issue shown in the gif with the delay only happening once, I think you can change the `gux-tooltip` styles to be like this:
```diff
:host {
  ...
  opacity: 0;
- transition: opacity 250ms ease-in-out;
- transition-delay: 750ms;
}

:host(.gux-show) {
  visibility: visible !important;
+ transition: opacity 250ms ease-in-out 750ms;
  opacity: 1;
}
```

daragh-king-genesys commented:
> I removed the setTimeout/clearTimeout for the css transition as I dont think it is needed.

If this change is not required to fix COMUI-1103 it should be moved to its own Issue and PR.

gavin-everett-genesys commented:
Utilising keyframes over transitions seems to have more flexibility and also prevents the issues that were happening with components that were using the `tooltip-title` which include `tag` and `tabs-advanced.` The delay on the animation is 1s and I added in `250ms` for the duration of the animation.

`animation-fill-mode` is a pretty cool property if you use forwards value on the property the target will retain the computed values set by the last keyframe which in this case is `opacity` of 1.

gavin-everett-genesys commented on src/components/stable/gux-tooltip/gux-tooltip.tsx:
```diff
@@ -81,13 +80,10 @@ export class GuxTooltip {
-    this.delayTimeout = setTimeout(() => {
-      this.isShown = true;
-    }, 750); // the css transition is 250ms
+    this.isShown = true;
```
This is needed to solve as it will delay the `gux-show` class which is responsible to animation.
